File: SDS_알고리즘/Day09/앱_7579.java
package SDS_알고리즘.Day09;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class 앱_7579 {
	static int N;
	static int M;
	static int totalCost;
	static int[] memmory;
	static int[] cost;
	static int[][] DP;

	public static void main(String[] args) throws Exception {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	    StringTokenizer st;

		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		memmory = new int[N + 1];
		cost = new int[N + 1];

		st = new StringTokenizer(br.readLine());
		for (int i = 1; i <= N; i++) {
			memmory[i] = Integer.parseInt(st.nextToken());
		}

		st = new StringTokenizer(br.readLine());
		for (int i = 1; i <= N; i++) {
			cost[i] = Integer.parseInt(st.nextToken());
			totalCost += cost[i];
		}
		DP = new int[N + 1][totalCost + 1];

		for (int app = 1; app <= N; app++) {
			int curCost = cost[app];
			int curMem = memmory[app];

			for (int c = 0; c <= totalCost; c++) {

				DP[app][c] = DP[app - 1][c];

				if (c < curCost) continue;

				if (DP[app][c] < DP[app - 1][c - curCost] + curMem) {
					DP[app][c] = DP[app - 1][c - curCost] + curMem;
				}
			}
		}

		for (int i = 1; i <= totalCost; i++) {
			if (DP[N][i] >= M) {
				System.out.println(i);
				break;
			}
		}
	}
}

File: 힣_코틀린_코테스터디/week6/앱_7579.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*


private var N = 0
private var M = 0
private var totalCost = 0
private lateinit var memory: IntArray
private lateinit var cost: IntArray
private lateinit var dp: Array<IntArray>

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    var st = StringTokenizer(br.readLine())

    N = st.nextToken().toInt()
    M = st.nextToken().toInt()
    memory = IntArray(N + 1)
    cost = IntArray(N + 1)

    st = StringTokenizer(br.readLine())
    for (i in 1..N) {
        memory[i] = st.nextToken().toInt()
    }

    st = StringTokenizer(br.readLine())
    for (i in 1..N) {
        cost[i] = st.nextToken().toInt()
        totalCost += cost[i]
    }
    dp = Array(N + 1) { IntArray(totalCost + 1) }

    for (app in 1..N) {
        val appCost = cost[app]
        val appMemory = memory[app]

        for (curCost in 0..totalCost) {
            dp[app][curCost] = dp[app - 1][curCost]

            if (curCost < appCost) continue
            if (dp[app][curCost] < dp[app - 1][curCost - appCost] + appMemory) {
                dp[app][curCost] = dp[app - 1][curCost - appCost] + appMemory
            }
        }
    }

    for (i in 1..totalCost) {
        if (dp[N][i] >= M) {
            println(i)
            break
        }
    }
}
